File: packages/mcp-resolver/src/commandSafety.ts
/**
 * Defense-in-depth structural validation, mirroring the checks performed by the
 * Kotlin resolver-core. A repository is untrusted: nothing it contains may ever
 * be assembled into a shell string.
 */

import { isAbsolute } from 'node:path';

const SHELL_METACHARACTERS = /[;&|`$<>(){}!\n\r]/;

export function isValidCommand(command: string): boolean {
  if (command.length === 0 || command.length > 1024) return false;
  if (SHELL_METACHARACTERS.test(command)) return false;
  // No quoting or substitution of any kind.
  if (command.includes("'") || command.includes('"')) return false;
  if (command.startsWith('..')) return false;

  if (isAbsolute(command)) {
    // Absolute paths are legitimate even when they contain spaces.
    return true;
  }

  // A relative command is either a single bare token (`node`, `npx`) or an
  // explicit `./<path>` into the working directory. Spaces are allowed only in
  // the explicit-relative form.
  const hasSeparator = command.includes('/') || command.includes('\\');
  if (hasSeparator) {
    if (!command.startsWith('./')) return false;
    return true;
  }
  return !/\s/u.test(command);
}

export function isValidArgs(args: string[]): boolean {
  if (args.length > 128) return false;
  return args.every(
    (arg) =>
      arg.length > 0 &&
      arg.length <= 4096 &&
      !SHELL_METACHARACTERS.test(arg) &&
      !arg.includes('\0'),
  );
}

export function validateResolved(resolved: {
  command: string;
  args: string[];
  workingDirectory: string | null;
}): void {
  if (!isValidCommand(resolved.command)) {
    throw new Error(`Invalid command: ${JSON.stringify(resolved.command)}`);
  }
  if (!isValidArgs(resolved.args)) {
    throw new Error(`Invalid args for ${JSON.stringify(resolved.command)}`);
  }
  if (resolved.workingDirectory !== null && resolved.workingDirectory.length === 0) {
    throw new Error('Working directory must be non-empty when provided');
  }
}

File: resolver/src/main/kotlin/com/creepyim/mcp/resolver/util/CommandSafety.kt
package com.creepyim.mcp.resolver.util

import java.nio.file.Path

/**
 * Structural validation for launch commands and arguments.
 *
 * A repository is untrusted: nothing it contains may ever be assembled into a
 * shell string. `command` must be a plain executable name or absolute path and
 * `args` must be a plain list. Any value that would require a shell to
 * interpret (metacharacters, pipes, redirection, substitution) is rejected.
 */
object CommandSafety {

    private val SHELL_METACHARACTERS = Regex("[;&|`\\$<>(){}!\\n\\r]")

    fun validateCommand(command: String, path: Path?): Boolean {
        if (command.isBlank() || command.length > 1024) return false
        if (SHELL_METACHARACTERS.containsMatchIn(command)) return false
        // A command is a single token — no spaces, no quoting, no substitution.
        if (command.any { it.isWhitespace() }) return false
        if (command.contains('\'') || command.contains('"')) return false
        if (command.startsWith("..")) return false
        if (command.contains('/')) {
            // Relative paths are only allowed as explicit ./<name> (never bare
            // traversal); otherwise the command must be absolute.
            if (!command.startsWith("./") && !Path.of(command).isAbsolute) return false
        }
        return true
    }

    fun validateArgs(args: List<String>): Boolean {
        if (args.size > 128) return false
        return args.none { arg ->
            arg.isEmpty() || arg.length > 4096 || SHELL_METACHARACTERS.containsMatchIn(arg)
        }
    }

    fun validateWorkingDirectory(dir: Path?, root: Path?): Boolean {
        if (dir == null) return true
        val normalized = dir.toAbsolutePath().normalize()
        if (!java.nio.file.Files.isDirectory(normalized)) return false
        if (root != null && !normalized.startsWith(root.toAbsolutePath().normalize())) return false
        return true
    }
}